blog/views.py

```diff
@@ -3,115 +3,136 @@
 - post_list - returns post list;
 """
 
-from django.shortcuts import render, get_object_or_404, redirect
+from django.shortcuts import get_object_or_404, redirect
 from .models import Post, Comment
 from django.utils import timezone
+from django.utils.decorators import method_decorator
+from django.core.urlresolvers import reverse_lazy
 from .forms import PostForm, CommentForm
 from django.contrib.auth.decorators import login_required
+from django.views.generic import View, ListView, DetailView, CreateView, UpdateView, DeleteView, TemplateView
 # Create your views here.
 
 
-def post_list(request):
-    """Show posts, which are published, ordered by published date."""
-    posts = (
-        Post.objects
-        .filter(published_date__lte=timezone.now())
-        .order_by('published_date')
-    )
-    return render(request, 'blog/post_list.html', {'posts': posts})
+class Protected(View):
+    """Deny access to view for unauthorized users."""
 
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Call base method with login_required extension."""
+        return super().dispatch(*args, **kwargs)
 
-def post_detail(request, pk):
-    """Show post details."""
-    post = get_object_or_404(Post, pk=pk)
-    return render(request, 'blog/post_detail.html', {'post': post})
 
+class PostListView(ListView):
+    """View for list of all published posts."""
 
-@login_required
-def post_new(request):
-    """Create new post."""
-    if request.method == 'POST':
+    context_object_name = 'posts'
+    template_name = 'blog/post_list.html'
+
+    def get_queryset(self):
+        """Return already published posts."""
+        return Post.objects.filter(published_date__lte=timezone.now()).order_by('published_date')
+
+
+class PostDetailView(DetailView):
+    """View with detail info about post."""
+
+    model = Post
+    template_name = 'blog/post_detail.html'
+
+
+class NewPost(CreateView, Protected):
+    """View for creating new post."""
+
+    model = Post
+    fields = ['title', 'text']
+    template_name = 'blog/post_edit.html'
+
+    def post(self, request, *args, **kwargs):
+        """Get author-field from request and save post to DB."""
         form = PostForm(request.POST)
         if form.is_valid():
             post = form.save(commit=False)
             post.author = request.user
             post.save()
```

Comment:
It's done automatically, you don't have to implement post method at all

Reply:
I quess, I have to use construction on 55-57 anyway - not in post, but in form_valid then.

Reply:
No, there should probably be a way to even encapsulate that into save method of form